packages/form-client/src/node/form-server.ts

```diff
@@ -1,78 +1,114 @@
 /********************************************************************************
  * Copyright (c) 2023 CrossBreeze.
  ********************************************************************************/
-import { injectable } from '@theia/core/shared/inversify';
+import { inject, injectable } from '@theia/core/shared/inversify';
 import * as net from 'net';
 import * as rpc from 'vscode-jsonrpc/node';
 import { FormEditorClient, FormEditorService } from '../common/form-client-protocol';
 
-import { CloseModel, CrossModelRoot, OpenModel, RequestModel, SaveModel, UpdateModel, OnSave } from '@crossbreeze/protocol';
-
-// socket connection, must match the one in model-server/launch.ts
-const SOCKET_OPTIONS = { port: 5999, host: 'localhost' };
+import { waitForTemporaryFileContent } from '@crossbreeze/core/lib/node';
+import {
+   CloseModel,
+   CrossModelRoot,
+   MODELSERVER_PORT_FILE,
+   OnSave,
+   OpenModel,
+   RequestModel,
+   SaveModel,
+   UpdateModel
+} from '@crossbreeze/protocol';
+import { URI } from '@theia/core';
+import { Deferred } from '@theia/core/lib/common/promise-util';
+import { WorkspaceServer } from '@theia/workspace/lib/common';
 
 /**
  * Backend service implementation that mainly forwards all requests from the Theia frontend to the model server exposed on a given socket.
  */
 @injectable()
 export class FormEditorServiceImpl implements FormEditorService {
-    protected initialized = false;
-    protected connection: rpc.MessageConnection;
-    protected client?: FormEditorClient;
+   protected initialized?: Deferred<void>;
+   protected connection: rpc.MessageConnection;
+   protected client?: FormEditorClient;
+
+   @inject(WorkspaceServer) protected workspaceServer: WorkspaceServer;
 
-    async initialize(): Promise<void> {
-        if (this.initialized) {
-            // ensure we only initialize RPC connection once
-            return;
-        }
-        const socket = new net.Socket();
-        const reader = new rpc.SocketMessageReader(socket);
-        const writer = new rpc.SocketMessageWriter(socket);
-        this.connection = rpc.createMessageConnection(reader, writer);
+   protected initialize(): Promise<void> {
+      if (this.initialized) {
```

Comment:
If the promise is rejected, will js see the this.initalized as false?

Reply:
We create the `initialized` object independent from the promise resolution. So as soon as we have that we always are initialized but this function will return the rejected promise as a result. So any calls that build on that, e.g., `open` etc. will also reject their call and should produce an error on the calling side.